app/master/cluster_master.py

```diff
@@ -279,7 +282,7 @@ def handle_result_reported_from_slave(self, slave_url, build_id, subjob_id, payl
         build = self._all_builds_by_id[int(build_id)]
         slave = self._all_slaves_by_url[slave_url]
         # If the build has been canceled, don't work on the next subjob.
-        if not build.is_finished:  # WIP(joey): This check should be internal to the Build object.
+        if not build.is_finished or build.has_error:  # WIP(joey): This check should be internal to the Build object.
```

Comment:
(Evan and I just talked about this in person)

My guess is that without it, not all of the slaves (especially the ones that didn't fail build_setup) will not be properly de-allocated from the build that they are working on unless this extra conditional statement is there.

It's important that ```scheduler.execute_next_subjob_or_free_executor(slave)``` gets called eventually on each slave so that the slave knows that it is free to work on new builds (and be added back to the free slave pool). I believe that in the event of a cancellation, those slaves will not be placed back into the free slave pool without this ```or build.has_error``` line.